edx/edx-enterprise
# ENT-296: Enable account-level consent requirement

**Description:**

At present, users are prompted to consent to the exact scope of consent that's relevant to the consent gate that they're attempting to pass - for example, if attempting to access a consent-gated courses, they'll be prompted for course consent; if attempting to access a consent-gated SSO provider, they'll be prompted for account consent.

This change adds a feature whereby, on a per-Enterprise Customer basis, a setting can be made that will force users to be prompted for account-level consent rather than course-specific consent. This change also patches the `EnterpriseCourseEnrollment` save method such that when a record is being saved for an EnterpriseCustomer with the above flag set, an account-wide consent audit record is written.

**JIRA:** [ENT-296](https://openedx.atlassian.net/browse/ENT-296)

**Dependencies:** None

**Sandbox:** https://ent296.opencraft.hosting/

**Merge deadline:** 18 June 2017

**Installation instructions:** No special installation steps; standard Enterprise installation

**Testing instructions:**

1. In [the Enterprise admin panel](https://ent296.opencraft.hosting/admin/enterprise/enterprisecustomer/fcd2b0e7-c687-4e19-af64-8889d835c15f/manage_learners), proxy-enroll a user in two courses
2. Ensure that the [EnterpriseCustomer is set to](https://ent296.opencraft.hosting/admin/enterprise/enterprisecustomer/fcd2b0e7-c687-4e19-af64-8889d835c15f/)
    1. require consent at enrollment
    2. *not* require account-level consent
3. As a user, browse to [one of](https://ent296.opencraft.hosting/courses/course-v1:OpenCraft+Secondary+Run1/courseware) [the courses](https://ent296.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/courseware) that you were enrolled in.
4. Observe the consent prompt; note that the messaging is specific to the single course you're attempting to access. Do not provide consent at this time.
5. Ensure that [EnterpriseCustomer is set to](https://ent296.opencraft.hosting/admin/enterprise/enterprisecustomer/fcd2b0e7-c687-4e19-af64-8889d835c15f/) require account-level consent.
6. Again, browse to one of the courses.
7. Observe that the messaging has changed, and is distinctly supportive of the premise that consent will be given for the entire account.
8. Provide consent.
9. Navigate to the dashboard and click on the _other_ course; note that you're taken directly to it without being prompted for consent.
10. In the Enterprise admin panel for the Enterprise Customer User; note that an account-wide consent audit log has been written, indicating global consent has been provided.

**Merge checklist:**

  **DO NOT** just add dependencies to `requirements/*.txt` files.

**Post merge:**
